src/plugins/ucx/ucx_backend.cpp

```diff
@@ -342,52 +348,75 @@ class nixlUcxBackendH : public nixlBackendReqH {
 void nixlUcxEngine::progressFunc()
 {
     using namespace nixlTime;
-    pthrActive = 1;
 
     vramApplyCtx();
 
+    {
+        std::unique_lock<std::mutex> lock(pthrActiveLock);
+        pthrActive = true;
+    }
+    pthrActiveCV.notify_one();
+
+    std::vector<pollfd> poll_fds;
+    for (auto &uw: uws) {
+        int fd;
+        if (ucp_worker_get_efd(uw->getWorker(), &fd) == UCS_OK)
+            poll_fds.push_back({fd, POLLIN, 0});
+        else
+            NIXL_ERROR << "Couldn't obtain fd for a worker";
+    }
+    poll_fds.push_back({pthrStopFds[0], POLLIN, 0});
+
+    std::vector<size_t> active;
+    for (size_t wid = 0; wid < uws.size(); wid++)
+        active.push_back(wid);
+
     while (!pthrStop) {
-        int i;
-        for(i = 0; i < noSyncIters; i++) {
-            uw->progress();
+        for (auto wid: active) {
+            bool made_progress = false;
+            ucs_status_t status = UCS_INPROGRESS;
+            const auto &uw = uws[wid];
+
+            while (status != UCS_OK) {
+                while (uw->progress())
+                    made_progress = true;
+
+                status = ucp_worker_arm(uw->getWorker());
+            }
+
+            if (made_progress && !wid)
+                notifProgress();
         }
-        notifProgress();
-        // TODO: once NIXL thread infrastructure is available - move it there!!!
-
-        // {
-        //     static uint64_t cnt = 0;
-        //     if ( !(cnt % 1000000)) {
-        //         std::cout << "Progress round" << std::endl;
-        //     }
-        //     cnt++;
-        // }
-
-        /* Wait for predefined number of */
-        us_t start = getUs();
-        while( (start + pthrDelay) > getUs()) {
-            std::this_thread::yield();
+
+        active.clear();
+        while (poll(poll_fds.data(), poll_fds.size(), -1) == 0);
+
+        for (size_t wid = 0; wid < uws.size() && wid < poll_fds.size(); wid++) {
+            if (poll_fds[wid].revents & POLLIN) {
+                active.push_back(wid);
```

Comment:
do we really have to maintain active() array? could just rearm what is needed in this loop
(could reset poll_fds[wid].revents to POLLIN before entering the loop to unite code)

Reply:
This PR became too big and disjointed. Removed the progress thread refactoring patch to be submitted separately.